### PR TITLE
change(state): Stop using iterators on column families with many deletions

### LINKED ISSUE #7618: Zebra full sync is taking at least 50% longer over the last 3 weeks
## Motivation

The scheduled full sync test timed out after 3 days. It usually takes just over 2 days.

The last time it ran successfully was 3 weeks ago.

There are two possible causes:
* mempool spam slowing down network communication
* we merged a PR that slowed down syncing in the last 3 weeks

### Logs

The failure is a timeout:
https://github.com/ZcashFoundation/zebra/actions/runs/6274009969/job/17038728970#step:9:21231

GitHub Actions workflow [CI Docker #6194](https://github.com/ZcashFoundation/zebra/actions/runs/6274009969) failed.

Event: schedule
Branch: [main](https://github.com/ZcashFoundation/zebra/tree/main)
Commit: [b737ccf5709b2905c7c53ce91eea17dd753a6c03](https://github.com/ZcashFoundation/zebra/commit/b737ccf5709b2905c7c53ce91eea17dd753a6c03)

<sup><i>Created by [jayqi/failed-build-issue-action](https://github.com/jayqi/failed-build-issue-action)</i></sup>

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

When we use a RocksDB iterator on a column family with lots of deleted keys, database performance decreases by up to 300x.

Closes #7618

### Specifications

This is a known issue with RocksDB, here are similar issues in other projects:
https://tracker.ceph.com/issues/55324
https://jira.mariadb.org/browse/MDEV-19670

## Solution

As a quick fix we can replace the iterator with `get(&key)`.
The permanent fix in PR #7392 stops deleting these keys entirely.

### Manual Testing

I did a local test of this fix, and Zebra is up to 1.2 million blocks in under 5 hours. This is faster than Marek's `main` branch checks on his machine, and about as fast as `v1.2.0` in CI:
https://github.com/ZcashFoundation/zebra/actions/runs/6049184682/job/16453440408

I think `v1.2.0` was about as fast on my machine, but I haven't done any detailed tests. (A manual check is ok, we can do detailed checks in #7649.)

## Review

This is an urgent fix which is blocking CI from working.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

Permanent fixes for all column families that delete keys:
- merge the permanent fix to these column families in PR #7392 
- #7664

Performance testing for this fix:
- #7649

### CODE REVIEW

File: zebra-state/src/service/finalized_state/disk_format/chain.rs
Comment (arya2) [inline zebra-state/src/service/finalized_state/disk_format/chain.rs]:
```suggestion
// We don't write empty history trees to disk, so we know this one is non-empty.
```

File: zebra-state/src/service/finalized_state/zebra_db/shielded.rs
Comment (arya2) [inline zebra-state/src/service/finalized_state/zebra_db/shielded.rs:126]: Could this reintroduce the panic bug? I don't think there should be a performance issue with reading the sprout tree from disk, it only happens once per run.

Comment (teor2345) [reply zebra-state/src/service/finalized_state/zebra_db/shielded.rs:126]: Yes, it could re-introduce the panic bug, but if the tree is only read once per run, it is unlikely that the panic bug will happen.

I've re-opened PR #7392, which contains a permanent fix for ~~this issue~~ the panic bug. I'll rebase it once this PR merges.

**Comment by teor2345** [comment]
Once we've verified this performance fix, I think we should admin-merge this PR, because it is currently blocked by CI failures in #7659 and #7661.

**Comment by teor2345** [comment]
### Manual Testing

I did a local test of this fix on my machine, and Zebra is up to 1.2 million blocks in under 5 hours. This is faster than Marek's `main` branch checks on his machine, and about as fast as `v1.2.0` in CI:
https://github.com/ZcashFoundation/zebra/actions/runs/6049184682/job/16453440408

I think `v1.2.0` was about as fast on my machine, but I haven't done any detailed tests. So it looks like this fixes the bug. (And the way we use iterators in other PRs doesn't cause similar bugs.)

But I'd appreciate someone else checking the performance as well. A manual check is ok, we can do detailed checks in #7649.

**Comment by upbqdn** [comment]
This PR is as fast as `v1.2.0`, so it resolves the slow sync.

I wonder if we could speed Zebra up even further by fixing a similar issue that degrades performance elsewhere, but we accepted it.

**Review by arya2 — APPROVED** [review]

**Comment by arya2** [comment]
In the merge queue, the test passed but the job failed:

> test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 45 filtered out; finished in 10465.70s
> tee: 'standard output': Broken pipe
> Error: Process completed with exit code 1.
> 
> https://github.com/ZcashFoundation/zebra/actions/runs/6405648679/job/17392070905#step:9:5457

**Comment by arya2** [comment]
@Mergifyio refresh

**Comment by teor2345** [comment]
> In the merge queue, the test passed but the job failed:
> 
> > test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 45 filtered out; finished in 10465.70s
> > tee: 'standard output': Broken pipe
> > Error: Process completed with exit code 1.
> > [ZcashFoundation/zebra/actions/runs/6405648679/job/17392070905#step:9:5457](https://github.com/ZcashFoundation/zebra/actions/runs/6405648679/job/17392070905#step:9:5457)

I reopened PR #7564 for this.

Since this PR is failing due to bug #7665, I'm going to admin-merge it.

**Pull request merged**
